### PR TITLE
Added articles for Security app's functions and React components

### PR DESCRIPTION
**Pull request by EmilK15**
… hasRole, SecureView and SecureRoute. Also added example of security usage for Menus and Routes.

### CODE REVIEW

**Comment by doitadrian** [comment]
## General
1. Please highlight all code with backticks. For example, it's not hasScope, but `hasScope`. Check this on all pages.

2. When submitting PRs, please write a complete description of what are you submitting. You don't have to write a whole trilogy as the initial issue did (😉), but at least a chunk of text with a beginning and an end. We are open source, we want the PRs (at least from team members) to look good and professional.

Note: I'll add a PR template and add some more guidelines into the CONTRIBUTING doc, just so other developers can have a clue of what we expect.

3. Please move the Plugins reference below the API/App sections, so it should be:
```
API
App
Plugins Reference
```

image

## `hasRole`
1. This should be "role", not "scope".
image

2. Minor addition
> After setting up a security role for a ...

I would just add "via the Roles form", just so the user knows where the "setting up a security role" is actually happening. Note: later we'll create a couple of simple pages that show the UI part of the app, and we'll link from here.

3. This sentence doesn't sound right to me.
> This can be used as a validator for your GraphQL queries/mutations.

Let's not use the word "validator", just so someone doesn't get confused with data validation. I would rather maybe reword the whole section, for example:
> After setting up a security role via the Roles form, you can utilize the `hasRole` helper function in order to determine whether a GraphQL field can be accessed by the client. 

4. After number 3, I would also explain and show what happens when a user doesn't have correct permissions ("Not authorized" error is shown in the response). You can get that by opening GraphQL playground and execute any query, e.g.:
```
{
  pageBuilder {
    listPages {
      data {
        id
      }
    }
  }
}
```

5.  Please add an empty line here:
image

6. In the 2nd code snippet, we're using `security-full-access`, which is actually a security group, not a role. Please set role there, e.g. `security-roles` (allows management of roles - complete CRUD is allowed).

7.  Similar to number 6, this is mentioning groups, it should be mentioning roles.

image

8. Let's call this GraphQL helpers, since these functions can only be used with GraphQL fields.
image

## `hasScope` 
The feedback I had for the `hasRole` sections applies here as well.

## `SecureRoute`
1. Just to be kinda more uniformed with the rest of the docs, for example...
image

... please remove the subtitles (When to use, How to import, Example), and try to make it a single chunk of text. Let's see how it's gonna look. If it turns out it's not good, we can revert.

2. Does't sound right:
> You can wrap your component with SecureRoute wherever your react-router routes are defined.

Something like:
> Use this component to prevent rendering a route if the user doesn't have sufficient permissions (security scopes or security roles). 

No need to mention `react-router` there, we don't wanna confuse the user. Basically, users don't even see the react-router directly (btw we have our own package `@webiny/react-router` for routing).

3. After the text in number 2., please start explaining the example that follows. You can use the quoted text you've added as a base, but please add a few words about the props that the `SecureRoute` component accepts (`roles` and `scopes) and to what they refer to (defined scopes in the code and created roles via the Roles form). And in general, describe what will happen if the user has the permissions, and what will happen if he doesn't. You've already mentioned that, but screenshots would be a nice addition as well.

We want the users to fully understand what they are doing and to what the presented content is referring to. We don't want them scratching their heads.

4. The quoted text does not need to be quoted, it's a regular text, explaining the current topic.

5. You've mentioned the "You are not authorized to view this route" message, but where does this come from? If you check the `SecureRoute` component, you'll see the text is actually loaded via the `secure-route-error` plugin. Which means, users can customize this, which again means this might be useful information. Please mention the plugin here, and also create a new entry [here ](https://deploy-preview-28--webiny-docs.netlify.com/docs/webiny-apps/security/development/plugin-reference/app). C/p the existing `security-scopes-list` section and customize it to your needs. Another example of a plugins reference page can be found [here](https://deploy-preview-28--webiny-docs.netlify.com/docs/webiny-apps/admin/development/plugins-reference/app).

## `SecureView`
1. Sentence incorrect:
image

It can be roles and scopes.

When I think about it more, I would reword the whole sentence. For example: "The`SecureView` component can be used to prevent rendering of a component in case the user doesn't have sufficient permissions (security roles and security groips).

2. Wrap the code with `function MyComponent() { ... }`, just so we are looking at a complete code snippet. Do this for both snippets (check other pages for this as well).

3. I would reword this:
> In the render method of the component you want to protect, you can wrap it with the SecureView component.

I would rather just say: "In the following example, the `YourComponent` component will be rendered only if the currently logged-in user has the `forms-editors` role.".

4. Try having a role and component name that are in the same context. For example, a role might be `security-roles`, and the component might be `<EditUserRoles/>`. This will make sense since we are showing the component only if the user has the `security-roles` role. The `forms-editors` and `YourComponent` example might suffice for some, but let's make sure it's clear for everybody.

5. Explicitly explain that the component can take `roles` and `scopes` props, but can also take a render prop which can give you more flexibility.

6. For the second example, explain what this approach is good for. It's good when you have a UI that consists of a few elements, and each element can be rendered depending on received values. Examine this example: `packages/app-page-builder/src/admin/plugins/menus.tsx:29`.

7. What does this mean? Can you provide an example? I don't think this is true also, am I wrong? I think the component either renders something, either returns `null`.
image

7. Check which feedback related to the `SecureRoute` section can be applied here as well.

**Comment by EmilK15** [comment]
The following has been added:
`hasScope`, `hasRole` GraphQL helper function documentation regarding use and error message display.
`SecureView`, `SecureRoute` component documentation regarding their use when creating a component that should only display given proper role and scope permissions.
Added reference definition for secure-route-error within App Plugins Reference.
Added screenshots relating to how to find `roles` and `scopes` within the Webiny Application itself.
Changed example for secureView to include names that were more relevant to `roles` and `scopes` defined.
Organized sidebars.js to include (API, App, Plugins Reference)
Changed React Component documentation for `SecureView` and `SecureRoute` to more closely resemble documentation for other React Components

**Comment by EmilK15** [comment]
Removed subtitles from `hasScope` and `hasRole` definition and moved error example to the bottom of the page for better readability.

**Comment by doitadrian** [comment]
### `hasRole`

1. The first few sentences are solid, but when we say that the helper can be used to determine.... the question is, **where** can we use it?

The answer is: in the `graphql-schema` plugins. We don't have this plugin explained yet, but basically, it is the plugin that you use when doing API development, in order to define your GQL schema / resolvers. The examples you've dug up all from these plugins as well.

So, I would add that here definitely. In the future, we'll have that explained in our `API Development` section. Hopefully soon :)

1.1. Please just change the `GraphQL field can be accessed by the client.` to `GraphQL field can be accessed by the user.` Not sure why I used the word `client` here. No point to use it if we're not using it in other places.

2. Continuing with the plugin.... the code looks good but cut off into three pieces like this, I think it might confuse the user. I would rather use one code block that contains everything. It will give a one-piece picture to the user, not a broken one :)

I made a simple but full example of a `graphql-schema` plugin, with the `hasScope` utilized:

```ts
import gql from "graphql-tag";
import { emptyResolver } from "@webiny/commodo-graphql";
import { hasScope } from "@webiny/api-security";

(...)

{
    type: "graphql-schema",
    name: "graphql-schema-my-app",
    schema: {
        typeDefs: gql`
            type Book {
                id: ID
                title: String
                pagesCount: Int
            }

            type BookInput {
                id: ID
                title: String
                pagesCount: Int
            }

            type MyAppQuery {
                getBook(id: ID): Book
            }

            type MyAppMutation {
                createBook(data: BookInput): Book
            }

            extend type Query {
                myApp: MyAppQuery
            }

            extend type Mutation {
                myApp: MyAppMutation
            }
        `,
        resolvers: {
            Query: {
                myApp: emptyResolver
            },
            Mutation: {
                myApp: emptyResolver
            },
            MyAppQuery: {
                getBook: () => { ... }
            },
            MyAppMutation: {
                createBook: () => { ... }
            }
        }
    },
    security: {
        shield: {
            MyAppQuery: {
                getBook: hasScope("my-app:books:get")
            },
            MyAppMutation: {
                createBook: hasScope("my-app:books:create")
            }
        }
    }
}

```

3. Note that in the above example, I changed some of the elements. More specifically, I don't use `security-roles` / `security-full-access`, `Security...` GQL types and resolvers in my examples. Showing security concepts on examples that also contain security-related terms will just confuse the user. We can use something else, and we should, just to make a clear distinction to the user - what is really part of the security code, and what is the user's custom code.

4. This is not a change request, but just a comment. I noticed we don't ever explain why the `security.shield` structure. I think we will cover that in a separate article, which will then contain links to this and `hasScope`.

5. I can't see what the error message says here, without opening the image in the new tab. Try reducing the width of the image (by reducing the size of the left side of the GQL playground, in which you've typed the query), it should make the text larger. 

image

### `hasScope`

1. This is not correct:
> After setting up a security scope for a specific user **via the Roles form**...

1.1 The same "client" word replacement. Sorry about that.

2. Make the code a single code block, as in `hasRole` section. BTW you can include comments if you want.

### `SecureRoute`
1. Missing code highlight in the first sentence (The **SecureRoute** component...).

2. This is not correct:
> The SecureRoute component can be used when you want certain routes to be available to specific roles

We are not setting just which roles are needed for the route to be visible, but also scopes. But all in all, I don's see the point of this sentence since you're explaining the same thing below the first code snippet (the import statement).

3. Missing image? The sentence does end with ":":
image

4. In the code snippet, as mentioned, please use roles / scopes that make more sense and more distinctly give an idea to the user what are the passed value. For `roles` prope, let's use e.g. `["my-app:books"]`, and for scopes maybe `["my-app:books:create", "my-app:books:read"]`.

5. A couple of things here:
image

(1) should be plural
(2) code highlight
(3) please just directly link the `secure-route-error`, and remove the "(which you can customize...)"
(4) Please use `"` instaed of `'`
(5) as mentioned in (3), please remove this

### `SecureView`

1. These are basically the same sentences here:
image

2. A couple of things here:
image

==> 0 Add a section title here, e.g. "Using render prop for selective rendering" or similar. More on this in the following text.

==> 1 This s not the point actually. The point is that when you use [render prop](https://reactjs.org/docs/render-props.html#using-props-other-than-render), you get check results as children function args, so you can granularly check what is the results of the roles checks and scopes checks, and with those results, selectively render different parts of the UI that's about to be rendered as the `SecureView` children.

==> 2 Change "In this case" to "In the following example"
==> 2.1 I'm not sure if you understand the keys "groups", "roles", and "users" in the `roles` prop. These props can be anything you like. You can name the property "apple" if you want. The point is, the keys you write here will enable you to access the results of the role check in the children function. So in your case, `groups` represent the result of the check: "does the user have a role `security-groups`". We can cover this additionally over slack if needed.

The same goes for `scopes` key. By writing `securityGroups: ["security-full-access"]`, you told the component to assign the result of the check "does the user have the `security-full-access`" to `securityGroups` property, which again, you can access in the `scopes` argument in the children function.

Note that this way of specifying the roles and scopes only makes sense when using render prop.

==> 3 `securityGroups` is not a good example of a role, and `security-full-access` even less so. Use rather something like ` booksCrud: ["my-app:books:crud"]`. Again, as discussed, an ideal example of a scope would be `my-app:books:create` or ``my-app:books:delete`, but for now let's use what we currently have in the system.

==> 4 Should `const { securityGroups } = groups;` in fact  be `const { securityGroups } = scopes;` ?

==> 5 This is too indented, should be moved to the left, below the `return`.

3. I don't understand this. Another benefit of what?
image

**Comment by EmilK15** [comment]
Added complete examples for hasRole and hasScope methods. Used e-library as an example for both hasRole and hasScope. Changed error image to appear clearer to the user for better readability.
Made full examples for SecureView and SecureRoute components. Made examples more relevant to e-library example. Added comments for more complex example for SecureView. 
Edited ending text for SecureView.

**Comment by EmilK15** [comment]
-Removed unnecessary comments from hasRole and hasScope examples.
-Added editBook in order to have the hasRole and hasScope examples match the SecureView and SecureRoute examples.
